enterprise/frontend/src/embedding-sdk/store/auth.ts

```diff
@@ -0,0 +1,159 @@
+  EmbeddingSessionToken,
+  FetchRequestTokenFn,
+  SDKConfig,
+} from "embedding-sdk";
+import type { SdkStoreState } from "embedding-sdk/store/types";
+import api from "metabase/lib/api";
+import { createAsyncThunk } from "metabase/lib/redux";
+import { refreshSiteSettings } from "metabase/redux/settings";
+import { refreshCurrentUser } from "metabase/redux/user";
+
+import { getOrRefreshSession } from "./reducer";
+import { getFetchRefreshTokenFn } from "./selectors";
+
+export const initAuth = createAsyncThunk(
+  "sdk/token/INIT_AUTH",
+  async (sdkConfig: SDKConfig, { dispatch }) => {
+    // Setup JWT or API key
+    const isValidJwtConfig =
+      sdkConfig.jwtProviderUri && sdkConfig.jwtProviderUri?.length > 0;
+    const isValidApiKeyConfig =
+      sdkConfig.apiKey && window.location.hostname === "localhost";
+
+    if (isValidJwtConfig) {
+      // JWT setup
+      api.onBeforeRequest = async () => {
+        const session = await dispatch(
+          getOrRefreshSession(sdkConfig.jwtProviderUri!),
+        ).unwrap();
+        if (session?.id) {
+          api.sessionToken = session.id;
+        }
+      };
+      // verify that the jwt is actually valid before proceeding
+      await dispatch(getOrRefreshSession(sdkConfig.jwtProviderUri!)).unwrap();
+    } else if (isValidApiKeyConfig) {
+      // API key setup
+      api.apiKey = sdkConfig.apiKey;
+    }
+    // Fetch user and site settings
+    const [user, siteSettings] = await Promise.all([
+      dispatch(refreshCurrentUser()),
+      dispatch(refreshSiteSettings({})),
+    ]);
+
+    if (!user.payload) {
+      // The refresh user thunk just returns null if it fails to fetch the user, it doesn't throw
+      const error = new Error(
+        "Failed to fetch the user, is the session valid?",
+      );
+      console.error(error);
+      throw error;
+    }
+    if (!siteSettings.payload) {
+      const error = new Error(
+        "Failed to fetch the site settings, is the session valid?",
+      );
+      console.error(error);
+      throw error;
+    }
+  },
+);
+
+export const refreshTokenAsync = createAsyncThunk(
+  "sdk/token/REFRESH_TOKEN",
+  async (url: string, { getState }): Promise<EmbeddingSessionToken | null> => {
+    // The SDK user can provide a custom function to refresh the token.
+    const customGetRefreshToken = getFetchRefreshTokenFn(
+      getState() as SdkStoreState,
+    );
+
+    const getRefreshToken = customGetRefreshToken ?? defaultGetRefreshTokenFn;
+
+    // # How does the error handling work?
+    // This is an async thunk, thunks _by design_ can fail and no error will be shown on the console (it's the reducer that should handle the reject action)
+    // The following lines are wrapped in a try/catch block that will catch any errors thrown, log them to the console as a big red errors, and re-throw them to make the thunk reject
+    // In this way we also support standard thrown Errors in the custom fetchRequestToken user provided function
+
```

Comment:
Not saying that the method is the best, but the only other way I thought of was doing the console.error in the reducer that listened to the `.rejected` action.
In theory reducers should be pure functions though, so I preferred doing it here (the stack should also be better in this way)